db_test.go

```diff
@@ -892,3 +892,52 @@ func expandSeriesSet(ss SeriesSet) ([]labels.Labels, error) {
 
 	return result, ss.Err()
 }
+
+func TestOverlappingBlocksDetectsAllOverlaps(t *testing.T) {
+	// Create 10 blocks that does not overlap (0-10, 10-20, ..., 100-110) but in reverse order to ensure our algorithm
+	// will handle that.
+	var metas = make([]BlockMeta, 11)
+	for i := 10; i >= 0; i-- {
+		metas[i] = BlockMeta{MinTime: int64(i * 10), MaxTime: int64((i + 1) * 10)}
+	}
+
+	testutil.Assert(t, len(OverlappingBlocks(metas)) == 0, "we found unexpected overlaps")
+
+	// Add overlapping blocks.
+
+	// o1 overlaps with 10-20.
+	o1 := BlockMeta{MinTime: 15, MaxTime: 17}
+	testutil.Equals(t, [][]BlockMeta{{metas[1], o1}}, OverlappingBlocks(append(metas, o1)))
+
+	// o2 overlaps with 20-30 and 30-40.
+	o2 := BlockMeta{MinTime: 21, MaxTime: 31}
+	testutil.Equals(t, [][]BlockMeta{{metas[2], o2}, {o2, metas[3]}}, OverlappingBlocks(append(metas, o2)))
+
+	// o3a and o3b overlaps with 30-40 and each other.
+	o3a := BlockMeta{MinTime: 33, MaxTime: 39}
+	o3b := BlockMeta{MinTime: 34, MaxTime: 36}
+	testutil.Equals(t, [][]BlockMeta{{metas[3], o3a, o3b}}, OverlappingBlocks(append(metas, o3a, o3b)))
```

Comment:
Hmm, a lot of complexity comes from achieving this, and I don't think there is any extra benefit to say:
`30-40` overlaps with `o3a` and `o3a` overlaps with `o3b`.

If we just check consecutive blocks, we simplify the algorithm a lot and still be able to catch overlapping blocks. I am not sure this complexity is worth it.

Reply:
It is worth it, because we want 100% accurate algorithm for it for potential "repair" flows. I would vote for leaving it accurate as much as possible. The complexity is not that bad now.

Reply:
There's generally also always some annoyance involved if you return "first match" in a sequence of errors. People have to go in circles then of fixing first match, retrying, getting next match, fix, ...

Reply:
Yep, instead of just returning the first match, we could return all the matches, I'm just thinking if returning:
`[{1, 2, 3}]` is any better than returning `[{1, 2}, {2, 3}]`. Also we would never be producing a third block if two are already overlapping.

All of the above code simplifies to:
```
prev := bm[0]
for _, b := range bm[1:] {
	if b.MinTime < prev.MaxTime {
		overlaps = append(overlaps, []BlockMeta{prev, b})
	}

	prev = b
}

return overlaps
```

Having said that I don't have a strong opinion here.

Reply:
I think your algorithm will not catch all the overlaps (for example `[1, 2, 3, 4]` where `2` is overlapping `1, 2 and 3`). That's the problem. IMO either we do "is there any overlap" or we do accurate overlapping groups.. Anything in the middle will confuse while debugging potential issues from logs.

Reply:
Basically mu rule of thumb is: what is more useful for debug readability and further implementations that will try to repair something?

IMO the grouped one, but I can see it's more complex

Reply:
Actually to full power of this algo is visible on extreme cases. We had bug in thanos that we have 200 blocks for same time range. This will end up as single group will all 200 of them for my proposed algo. 

If we end up doing just pairs, we will see 200^2 pairs that will not help really to understand the trend/problem